calico-vpp-agent/watchers/net_watcher.go

```diff
@@ -0,0 +1,210 @@
+// Copyright (C) 2021 Cisco Systems Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watchers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/projectcalico/vpp-dataplane/calico-vpp-agent/common"
+	networkv3 "github.com/projectcalico/vpp-dataplane/calico-vpp-agent/network"
+	"github.com/projectcalico/vpp-dataplane/vpplink"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/tomb.v2"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type VRF struct {
+	Tables [2]uint32 // one for ipv4, one for ipv6
+}
+
+type NetworkDefinition struct {
+	VRF               VRF
+	Vni               uint32
+	Name              string
+	LoopbackSwIfIndex uint32
+	Range             string
+}
+
+type NetWatcher struct {
+	log                *logrus.Entry
+	vpp                *vpplink.VppLink
+	client             client.WithWatch
+	stop               chan struct{}
+	networkDefinitions map[string]*NetworkDefinition
+}
+
+func NewNetWatcher(vpp *vpplink.VppLink, log *logrus.Entry) *NetWatcher {
+	kubernetesClient, err := NewClient(10 * time.Second, []func(s *runtime.Scheme) error{networkv3.AddToScheme})
+	if err != nil {
+		panic(fmt.Errorf("failed instantiating kubernetes client: %v", err))
+	}
+	w := NetWatcher{
+		log:                log,
+		vpp:                vpp,
+		client:             *kubernetesClient,
+		stop:               make(chan struct{}),
+		networkDefinitions: make(map[string]*NetworkDefinition),
+	}
+	return &w
+}
+
+func (w *NetWatcher) WatchNetworks(t *tomb.Tomb) error {
+	w.log.Infof("Net watcher starts")
+	netList := &networkv3.NetworkList{}
+	err := w.client.List(context.Background(), netList, &client.ListOptions{})
+	if err != nil {
+		return err
+	}
+	networks := make(map[string]*NetworkDefinition)
+	for _, net := range netList.Items {
+		netDef, err := w.OnNetAdded(&net)
+		if err != nil {
+			w.log.Error(err)
+			return err
+		}
+		networks[netDef.Name] = netDef
```

Comment:
An potential option in order not to send the whole map over the chan would be to pass the networks sperately, and use the `NetsSynced` event just as a signaling mechanism.
````
for _, net := range netList.Items {
        ...
        // send an event for each added network in the loop
        common.SendEvent(common.CalicoVppEvent{
		Type: common.NetAdded,
		New:  netDef.Copy(),
	})
       ...
}
// And send a NetsSynced event just to signal that everything is in sync
common.SendEvent(common.CalicoVppEvent{Type: common.NetsSynced})
````

Reply:
When I send netsSynced to the cni server it needs to make sure it knows all networks and sync its map. I think the way you are suggesting does not insure that all `NetAdded` events have arrived and all networks are synced in cni server at the moment `netsSynced` arrives.

Reply:
or are we sure given that it is the same channel so when netSynced arrives all previous messages have already arrived?

Reply:
Yes, the channel guarantees the ordering, so yes I think we can rely on this to be sure that when `netSynced` arrives, all the previous messages have been handled.
The thing that worries me a bit with passing the map over the chan is that it's fairly easy to forget about it in the future & start messing with the same struct in both places